statsmodels/survey/summary_stats.py

```diff
@@ -0,0 +1,538 @@
+        """
+        Minor error checking to make sure user supplied any of
+        strata, cluster, or weights. For unspecified subgroup labels
+        an array of ones is created
+
+        Parameters
+        ----------
+        strata : array-like or None
+            Strata for each observation. If none, an array
+            of ones is constructed
+        cluster : array-like or None
+            Cluster for each observation. If none, an array
+            of ones is constructed
+        weights : array-like or None
+            The weight for each observation. If none, an array
+            of ones is constructed
+
+        Returns
+        -------
+        vals[0] : ndarray
+            array of the strata labels
+        vals[1] : ndarray
+            array of the cluster labels
+        vals[2] : ndarray
+            array of the observation weights
+        """
+        if all([x is None for x in (strata, cluster, weights)]):
+            raise ValueError("""At least one of strata, cluster, and weights
+                             musts not be None""")
+        v = [len(x) for x in (strata, cluster, weights) if x is not None]
+        if len(set(v)) != 1:
+            raise ValueError("""lengths of strata, cluster, and weights
+                             are not compatible""")
+        n = v[0]
+        vals = []
+        for x in (strata, cluster, weights):
+            if x is None:
+                vals.append(np.ones(n))
+            else:
+                vals.append(np.asarray(x))
+
+        return vals[0], vals[1], vals[2]
+
+
+class SurveyStat(object):
+    """
+    Estimation and inference for summary statistics in complex surveys.
+
+    Parameters
+    -------
+    design : SurveyDesign object
+
+    Attributes
+    ----------
+    est : ndarray
+        The point estimates of the statistic, calculated on the columns
+        of data.
+    vc : ndarray
+        The variance-covariance of the estimates.
+    pseudo : ndarray
+        The jackknife pseudo-values.
+    """
+
+    def __init__(self, design, mse):
+        self.design = design
+        self.mse = mse
+
+    def bootstrap(self, replicates, index=None):
+        """
+        Calculates bootstrap standard errors
+
+        Parameters
+        ----------
+        stat : object
+            Object of class SurveyMean, SurveyTotal, SurveyPercentile, etc
+        replicates : integer
+            The number of replicates that the user wishes to specify
+
+        Returns
+        -------
+        est : ndarray
+            The point estimates of the statistic, calculated on the columns
+            of data.
+        vc : ndarray
+            The variance-covariance of the estimates.
+        """
+        if index is not None:
+            est = self._stat(self.design.weights, index)
+        else:
+            est = self._stat(self.design.weights)
+
+        jdata = []
+        has_reps = self.design.rep_weights is not None
+
+        if has_reps:
```

Comment:
Too much code duplication here between the `has_reps=True` and `has_reps=False` branches.  Refactor so that common steps are not repeated in the source.